src/octoprint/util/yaml_util.py

```diff
@@ -0,0 +1,76 @@
+# -*- coding: utf-8 -*-
+import io
+
+
+def safe_load(file=None, path=None):
+    """
+    Safely and performantly loads yaml data from the given source. Either a
+    path or a file must be passed in.
+
+    :rtype: Union[dict[Hashable, Any], list, None]
+    :type path: str | None
+    :type file: typing.TextIO | None
+    """
+    if path is not None:
+        assert file is None
+        with io.open(path, "rt", encoding="utf-8") as f:
+            return safe_load(file=f)
+
+    if file is not None:
+        assert path is None
+
+    assert (path is not None) or (file is not None)
+
+    import yaml
+
+    try:
+        from yaml import CSafeLoader as SafeLoader
+    except ImportError:
+        from yaml import SafeLoader
+
+    return yaml.load(file, Loader=SafeLoader)
+
+
+def safe_dump(data, file=None, path=None, **kwargs):
+    """
+    Safely and performantly dumps the `data` to yaml. If both file and path
+    are None, the value is returned as a string.
+
+    :rtype: str | None
+    :type data: object
+    :type path: str | None
+    :type file: typing.TextIO | None
+    """
+    if path is not None:
+        assert file is None
+        with io.open(path, "wt", encoding="utf-8") as f:
```

Comment:
I was hoping that this could be solved through user education, although it probably is less hassle to do it in code :(

Looks like even python 2.7 supports `utf-8-sig`, which deals with Notepad's dumb little thing. Is there a reason `bom_aware_open` was written instead of using that? Also, is there a particular reason `bom_aware_open`'s default encoding is ascii?

If not, then I'd love to delete that code in my next PR :grin:

Reply:
> I was hoping that this could be solved through user education, although it probably is less hassle to do it in code :(

User education is a noble goal that is also my preferred go-to, but simply fails in the field from my experience more often than not. And here Bad Stuff can be prevented easily with a bit of code, so it's frankly not worth the frustration of constantly seeing users run into this issue and trying to educate them on it again and again and again when all it takes is a few lines of code to solve it transparently. There are other battles to be fought that are worth more.

> Looks like even python 2.7 supports `utf-8-sig`, which deals with Notepad's dumb little thing. Is there a reason `bom_aware_open` was written instead of using that?

IIRC (and mind you, the introduction of `bom_aware_open` was over six years ago so my memory is a bit fuzzy), simply using `utf-8-sig` didn't work at the time because it would then *require* the presence of the BOM and thus fail on files without it, which is why I went through the hassle of first detecting if the `-sig` encoding was needed before using it (which is pretty much all that `bom_aware_open` does). However, I currently cannot reproduce this with my local 2.7 environment anymore, so either I'm misremembering things here, or this behaviour back then was actually buggy and got fixed since, or behaviour simply got changed when switching from `codecs.open` to `io.open` during the big Python 3 compatibility refactoring.

Considering that it no longer seems to be an issue, I guess `bom_aware_open` can be retired (though since there might be plugins out there using it it needs a deprecation annotation and then turned into a transparent wrapper around `io.open` with `-sig` attachment to the `encoding` value as needed).

> Also, is there a particular reason `bom_aware_open`'s default encoding is ascii?

IIRC, back then `open` calls defaulted to ascii, so I did too.

Reply:
resolved for https://github.com/OctoPrint/OctoPrint/issues/4228